src/dbnode/storage/namespace.go

```diff
@@ -365,11 +365,23 @@ func newDatabaseNamespace(
 	return n, nil
 }
 
-// SetSchema implements namespace.SchemaListener.
-func (n *dbNamespace) SetSchema(value namespace.SchemaDescr) {
+// SetSchemaHistory implements namespace.SchemaListener.
+func (n *dbNamespace) SetSchemaHistory(value namespace.SchemaHistory) {
 	n.Lock()
-	n.schemaDescr = value
-	n.Unlock()
+	defer n.Unlock()
+
+	metadata, err := namespace.NewMetadata(n.ID(), n.nopts.SetSchemaHistory(value))
+	if err != nil {
+		n.log.Error("can not update namespace metadata with empty schema history", zap.Stringer("namespace", n.ID()), zap.Error(err))
+		return
+	}
+
+	if schema, ok := value.GetLatest(); ok {
```

Comment:
Could you move this check to the top and just do an early return if the input is invalid? I.E:

```golang
n.Lock()
defer n.Unlock()

schema, ok := value.GetLatest()
if !ok {
    // error log
    return
}

metadata, err := ...
if err != nil {
    // log error
    return
}

n.schemaDescr = ...
n.metadata = ...
```

Reply:
The previous block is also a sanity check and early return. Only after both blocks are valid, will we do an assignment to member var.

Reply:
yeah I know but I think validating the input (the history) first before trying to make the metadata makes more sense. Also it reduces the nesting and just makes the logic easier to follow if its a series of guard clauses followed by doing the actual update with no elses